src/Node.hs

```diff
@@ -493,7 +493,7 @@ recvNext packing limit (LL.ChannelIn channel) = do
             when (remaining <= 0) (throw LimitExceeded)
             mbs <- Channel.readChannel channel
             case mbs of
-                Nothing -> return (BS.empty, End)
+                Nothing -> go remaining (next Nothing)
                 Just bs ->
                     let !remaining' = remaining - BS.length bs
                     in  go remaining' (next (Just bs))
```

Comment:
Not directly related to the fix, and just as a style recommendation, I would put the ! on the `go !remaining` argument, rather than using `let !` at the call site, ie

    go !remaining decoder = ...
        ....
             case mbs of
                 Nothing -> go remaining  (next Nothing)
                 Just bs -> go remaining' (next (Just bs))
                     where remaining' = remaining - BS.length bs

The advantage is that we only have to put the ! at the one definition site rather than the potentially multiple call sites. It also means when reading we can see immediately that the `remaining` is strict rather than having to scan the whole body. It also means there's nothing to think about if refactoring to inline or un-inline a let/where clause, e.g.

                 Just bs -> go (remaining - BS.length bs) (next (Just bs))